### PR TITLE
HDDS-5523. Fix multipart upload failure in s3 compatibility tests

### PR DESCRIPTION
**Pull request by GeorgeJahad**
## What changes were proposed in this pull request?

When running the [s3 compatibility tests](https://github.com/apache/ozone/tree/master/hadoop-ozone/dist/src/main/smoketest/s3/s3_compatbility_check.sh), I noticed AWS ignores invalid date strings in the copy-source-parameters, (copySourceIfModifiedSince, and copySourceIfUnmodifiedSince) of the upload-part-copy cmd.

Date strings in the future are also considered invalid.

I was unable to find anything that documented this behaviour explicitly.  I'm just going on the results from the compatibility test.

@kerneltime  mentioned that rfc2616 describes the behaviour for a similiar header parameter: 'If-Modified-Since' in section [14.25]( https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.25)

That seems to be what AWS is basing their implementation on.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5523

## How was this patch tested?

Ran the s3 compatibility tests against aws as well as s3 gateway.

### CODE REVIEW

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 1]
```diff
@@ -858,15 +860,29 @@ private static int parsePartNumberMarker(String partNumberMarker) {
     return partMarker;
   }
 
-  private static long parseOzoneDate(String ozoneDateStr) throws OS3Exception {
+  // If the return is empty, the precondition should be ignored
```
Comment (kerneltime) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:863]: Can you please update this comment and others in the method.
```suggestion
  // Parses date string and return long representation. Returns an empty if DateStr is null or invalid. Date in the future are considered invalid.
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 2]
```diff
@@ -858,15 +860,29 @@ private static int parsePartNumberMarker(String partNumberMarker) {
...
+  private static OptionalLong parseOzoneDate(String ozoneDateStr)
```
Comment (kerneltime) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:864]:
```suggestion
  private static OptionalLong parseAndValidateDate(String ozoneDateStr)
```

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java [hunk 3]
```diff
@@ -208,6 +218,73 @@ public void testMultipartIfModifiedSince() throws Exception {
     }
   }
 
+  // Test operations with 2 datestrings; one valid, the other invalid;
+  // confirm the invalid one is ignored and the valid precondition tested
```
Comment (kerneltime) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java:222]: Thank you for adding this, we should have 12 test cases or more, 4 based on true and false condition for timestamps that are valid, 4 with either one of the time stamps is null and the other true or false, 4 for ether timestamp in the future and other time stamp true or false but valid and finally actual bad input.

Comment (GeorgeJahad) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java:222]: Will do, Thanks @kerneltime!

Comment (GeorgeJahad) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java]: hey @kerneltime I've added the tests you suggested to s3/endpoint/TestMultipartUploadWithCopy.java.  There ended up being 25 different combinations total.  I had to restructure the tests a bit to reduce the repeated boilerplate.  Hopefully those changes won't be too hard to follow.


Here are just the changes since your review, if that makes it easier:
https://github.com/GeorgeJahad/ozone/compare/3e7edd84a1a7cbf3e0c8a95c05ae971be171ffa1..f567f523bd12f88ff6db96c5cfdfa6d0a853399a#diff-bd6abd262ed8e24eaeea5cb0e7c2a1a128411b284f9f27dae1ad47be6717a940

Thanks again!

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java [hunk 4]
```diff
@@ -156,56 +172,104 @@ public void testMultipart() throws Exception {
     }
   }
 
+  /* The next two tests excercise all the combinations of modification times.
```
Comment (kerneltime) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java:175]:
```suggestion
  /* The next two tests exercise all the combinations of modification times.
```

**Comment by kerneltime** [comment]
The changes looks complete now, have you tried running this against AWS S3 to see if the responses match in all cases?

**Comment by kerneltime** [comment]
I would recommend something like this to better explain all the variations and possible outcomes.

``` 
// CopyIfTimestampTestCase captures all the possibilities for the time stamps 
  // that can be passed into the multipart copy with copy-if flags for 
  // timestamps. Only some of the cases are valid others should raise an 
  // exception.
  // Time stamps can be, 
  // 1. after the timestamp on the object but still a valid time stamp 
  // (in regard to wall clock time on server)
  // 2. before the timestamp on the object
  // 3. In the Future beyond the wall clock time on the server
  // 4. Null
  // 5. Unparsable 
  // 
  public enum CopyIfTimestampTestCase {
    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_FUTURE_TS(afterSourceKeyModificationTimeStr, afterSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_PAST_TS(afterSourceKeyModificationTimeStr, beforeSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_NULL(afterSourceKeyModificationTimeStr, null, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_ABSOLUTE_FUTURE(afterSourceKeyModificationTimeStr, futureTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_FUTURE_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(afterSourceKeyModificationTimeStr, UNPARSABLE_TIME_STR, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_PAST_TS_UNMODIFIED_SINCE_FUTURE_TS(beforeSourceKeyModificationTimeStr, afterSourceKeyModificationTimeStr, null),
    MODIFIED_SINCE_PAST_TS_UNMODIFIED_SINCE_PAST_TS(beforeSourceKeyModificationTimeStr, beforeSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_PAST_TS_UNMODIFIED_SINCE_NULL(beforeSourceKeyModificationTimeStr, null, null),
    MODIFIED_SINCE_PAST_TS_UNMOFIFIED_SINCE_ABSOLUTE_FUTURE(beforeSourceKeyModificationTimeStr,futureTimeStr, null),
    MODIFIED_SINCE_PAST_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(beforeSourceKeyModificationTimeStr, UNPARSABLE_TIME_STR, null),
    MODIFIED_SINCE_NULL_TS_UNMODIFIED_SINCE_FUTURE_TS(null, afterSourceKeyModificationTimeStr, null),
    MODIFIED_SINCE_NULL_TS_UNMODIFIED_SINCE_PAST_TS(null, beforeSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_NULL_TS_UNMODIFIED_SINCE_NULL_TS(null, null, null),
    MODIFIED_SINCE_NULL_TS_UNMODIFIED_SINCE_ABSOLUTE_TS(null, futureTimeStr, null),
    MODIFIED_SINCE_NULL_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(null, UNPARSABLE_TIME_STR, null),
    MODIFIED_SINCE_UNPARSABLE_TS_UNMODIFIED_SINCE_FUTURE_TS(UNPARSABLE_TIME_STR, afterSourceKeyModificationTimeStr, null),
    MODIFIED_SINCE_UNPARSABLE_TS_UNMODIFIED_SINCE_PAST_TS(UNPARSABLE_TIME_STR, beforeSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_UNPARSABLE_TS_UNMODIFIED_SINCE_NULL_TS(UNPARSABLE_TIME_STR, null, null),
    MODIFIED_SINCE_UNPARSABLE_TS_UNMODIFIED_SINCE_ABSOLUTE_TS(UNPARSABLE_TIME_STR, futureTimeStr, null),
    MODIFIED_SINCE_UNPARSABLE_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(UNPARSABLE_TIME_STR, UNPARSABLE_TIME_STR, null),
    MODIFIED_SINCE_ABS_FUTURE_TS_UNMODIFIED_SINCE_FUTURE_TS(futureTimeStr, afterSourceKeyModificationTimeStr, null),
    MODIFIED_SINCE_ABS_FUTURE_TS_UNMODIFIED_SINCE_PAST_TS(futureTimeStr, beforeSourceKeyModificationTimeStr, S3ErrorTable.PRECOND_FAILED.getCode()),
    MODIFIED_SINCE_ABS_FUTURE_TS_UNMODIFIED_SINCE_NULL_TS(futureTimeStr, null, null),
    MODIFIED_SINCE_ABS_FUTURE_TS_UNMODIFIED_SINCE_ABSOLUTE_TS(futureTimeStr, futureTimeStr, null),
    MODIFIED_SINCE_ABS_FUTURE_TS_UNMODIFIED_SINCE_UNPARSABLE_TS(futureTimeStr, UNPARSABLE_TIME_STR, null);
    private String modifiedTimestamp;
    private String unmodifiedTimestamp;
    private String errorCode;
    CopyIfTimestampTestCase(String modifiedTimestamp, String unmodifiedTimestamp, String errorCode) {
      this.modifiedTimestamp = modifiedTimestamp;
      this.unmodifiedTimestamp = unmodifiedTimestamp;
      this.errorCode = errorCode;
    }

    @Override
    public String toString() {
      return " Modified:" + this.modifiedTimestamp
          + " Unmodified:" + this.unmodifiedTimestamp
          + " ErrorCode:" + this.errorCode;
    }
  }
  @Test
  public void testMultipartTSHeaders() throws Exception {
    for (CopyIfTimestampTestCase t : CopyIfTimestampTestCase.values() ) {
      try {
        uploadPartWithCopy(t.modifiedTimestamp, t.unmodifiedTimestamp);
        if (t.errorCode != null) {
          fail("Fail test:" + t);
        }
      } catch (OS3Exception ex) {
        if (t.errorCode == null) {
          fail("Failed test:" + t );
        }
      }
    }
  }
```

**Comment by GeorgeJahad** [comment]
Hmmm, I like that enum approach a lot.  Let me play with it a bit, and make sure I don't see any problems.

(Also, thank you for typing it all out.  Totally beyond the call of duty.)

**Comment by GeorgeJahad** [comment]
> The changes looks complete now, have you tried running this against
>  AWS S3 to see if the responses match in all cases?

yes

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java [hunk 5]
```diff
@@ -172,103 +173,132 @@ public void testMultipart() throws Exception {
...
       try {
-        uploadPartWithCopy(ts, beforeSourceKeyModificationTimeStr);
-        fail("testMultipartIfModifiedIsTrueOrInvalid");
+        uploadPartWithCopy(t.modifiedTimestamp, t.unmodifiedTimestamp);
+        if (t.errorCode != null) {
+          fail("Fail test:" + t);
+        }
       } catch (OS3Exception ex) {
-        assertEquals(ex.getCode(), S3ErrorTable.PRECOND_FAILED.getCode());
+        if (t.errorCode == null) {
```
Comment (kerneltime) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java:298]: Should this check the ERROR_CODE explicitly?

Comment (GeorgeJahad) [reply hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java]: yes, thanks @kerneltime!.  fixed now.

**Review by kerneltime — APPROVED** [review]

**Comment by GeorgeJahad** [comment]
Thanks @kerneltime!

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 6]
```diff
@@ -858,15 +860,29 @@ private static int parsePartNumberMarker(String partNumberMarker) {
...
     try {
       ozoneDateInMs = OzoneUtils.formatDate(ozoneDateStr);
     } catch (ParseException e) {
-      throw S3ErrorTable.newError(S3ErrorTable
-          .INVALID_ARGUMENT, ozoneDateStr);
+      // if time not parseable, then return empty()
+      return OptionalLong.empty();
+    }
+
+    long currentDate = new Date().getTime();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:879]: I think we can avoid creating a new `Date` object:

```suggestion
    long currentDate = System.currentTimeMillis();
```

File: hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java [hunk 7]
```diff
@@ -89,6 +95,26 @@ public static void setUp() throws Exception {
...
+    afterSourceKeyModificationTimeStr =
+        OzoneUtils.formatTime(sourceKeyLastModificationTime + DELAY_MS);
+    futureTimeStr =
+        OzoneUtils.formatTime(sourceKeyLastModificationTime +
+            1000 * 60 * 24);
+
+    // Make sure DELAY_MS has passed, otherwise
+    //  afterSourceKeyModificationTimeStr will be in the future
+    //  and thus invalid
+    long currentTime = new Date().getTime();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestMultipartUploadWithCopy.java:113]: Same here:

```suggestion
    long currentTime = System.currentTimeMillis();
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 8]
```diff
@@ -858,15 +860,29 @@ private static int parsePartNumberMarker(String partNumberMarker) {
     return partMarker;
   }
 
-  private static long parseOzoneDate(String ozoneDateStr) throws OS3Exception {
+  // Parses date string and return long representation. Returns an
+  // empty if DateStr is null or invalid. Dates in the future are
+  // considered invalid.
+  private static OptionalLong parseAndValidateDate(String ozoneDateStr)
+      throws OS3Exception {
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:867]: No longer thrown:

```suggestion
  private static OptionalLong parseAndValidateDate(String ozoneDateStr) {
```

(Also in `checkCopySourceModificationTime`.)

**Review by adoroszlai — COMMENTED** [review]
Thanks @GeorgeJahad for the patch.  I would like to suggest some minor improvements.

**Review by adoroszlai — APPROVED** [review]
Thanks @GeorgeJahad for updating the patch.

**Comment by adoroszlai** [comment]
Thanks @GeorgeJahad for the patch, @kerneltime for the review.
